tokio/src/runtime/task/trace/mod.rs

```diff
@@ -317,20 +319,28 @@ cfg_rt_multi_thread! {
         drop(synced);
 
         // notify each task
-        let mut traces = vec![];
+        let mut tasks = vec![];
         owned.for_each(|task| {
             // set the notified bit
             task.as_raw().state().transition_to_notified_for_tracing();
+            // store the raw tasks into a vec
+            tasks.push(task.as_raw());
```

Comment:
We should be taking a refcount to the task instead of using `as_raw`.

Reply:
What's the API for that?

Reply:
You can add this:
```rs
impl<S: 'static> Clone for Task<S> {
    fn clone(&self) -> Task<S> {
        self.raw.ref_inc();
        Task::new(self.raw)
    }
}
```

Reply:
Done!

Reply:
Huh. I realized that `transition_to_notified_for_tracing` actually performs a refcount increment. Because of that, you don't need the additional refcount increment. However, I would probably prefer to see this refcount management a bit more wrapped up in a less error-prone api.

Reply:
Hm. Unfortunately, now I'm quite confused by how refcount management is supposed to work. I removed only the increment in `transition_to_notified_for_tracing`, but doing so causes the tests to SIGABRT:
```
free(): invalid pointer
```

What's the relationship between notifying a task, polling a task, and its refcount?

Reply:
The behavior of each function is documented in [`harness.rs`](https://github.com/tokio-rs/tokio/blob/master/tokio/src/runtime/task/harness.rs). There's also a bunch of documentation at the top of [`mod.rs`](https://github.com/tokio-rs/tokio/blob/master/tokio/src/runtime/task/mod.rs).

Basically, each `Task` (or `Notified`) object owns a refcount. Various methods will consume the refcount. We want the refcount that `transition_to_notified_for_tracing` creates to be owned by a `Task` somehow, rather than implicitly via a `RawTask`. This is because it generally makes code much more clear, and it also makes the code more robust towards error paths due to RAII cleanup.

Reply:
Got it, thanks! So, to summarize, the crash occurred because:
- cloning `Task` incremented the refcount
- polling the task decremented the refcount
- dropping `Task` decremented the refcount

...which is one too many decrements. `transition_to_notified_for_tracing` needs to increment the refcount (much like `transition_to_notified_by_ref`), because polling is going to consume that count.